### ISSUE TITLE
Propagate promotions out of cast patterns

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/language#3099]
Currently, promotions that occur inside the subpattern of a cast pattern have no effect outside the cast pattern. For example:
```dart
test(Object x) {
  if (x case int() as num && var y) {
    // `x` and `y` both have static type `num`, not `int`.
  }
}
```

The reason this happens is because flow analysis doesn't understand that the subpattern of a cast pattern is matching the same value as the cast pattern itself. In effect, the code above is analyzed in the same way as:
```dart
test(Object x) {
  if (x case Object(asNum: int()) && var y) {
    // `x` and `y` both have static type `num`, not `int`.
  }
}

extension on Object {
  num get asNum => this as num;
}
```

As @lrhn points out in https://dart-review.googlesource.com/c/sdk/+/304320/comment/17d98d6d_da9be7a5/, it would make more sense for the promotion to `int` to propagate from the subpattern to the outer pattern, so that in the first example above, both `x` and `y` would be promoted to `int`.

**Comment by stereotype441** [comment]
I think the best way to do this would be using a similar technique to how records get promoted based on the subpatterns that match their fields: while visiting the subpattern, assume the matched value is stored in a synthetic variable of type `X` (where `X` is the cast type); once the subpattern match completes, examine the promoted type of the synthetic variable, and attempt to promote the outer matched value (and the scrutinee, if applicable) based on that promoted type.

So for example:
```dart
class A {}
class B implements A {}
class C implements B {}
class D implements A {}
test1(A x) {
  if (x case (C() && var y) as B && var z) {
    // While visiting the subpattern `C()`, the matched value is assumed to be
    // held in a synthetic variable of type `B`. After visiting `C()`, that
    // synthetic variable's type has been promoted to `C`, so `y` has type
    // `C`. When exiting the cast pattern, since the synthetic variable has been
    // promoted to `C`, we to attempt to promote `x` to `C`. This promotion
    // succeeds, since `C <: A`. So `x` is promoted to `C`, and `z` has type
    // `C`.
  }
}
test2(D x) {
  if (x case (C() && var y) as B && var z) {
    // As above, except that when we try to promote `x` to `C`, the promotion
    // fails, because `C` is not a subtype of `D`. So `y` has type `C`, but `x`
    // and `z` have type `D`.
  }
}
```

**Comment by stereotype441** [comment]
CC @dart-lang/language-team

**Comment by munificent** [comment]
Yeah, this seems like a reasonable change to me. Though I do wonder how often in comes into play in real world code.

**Comment by lrhn** [comment]
I don't expect cast/null-bang patterns to occur that often in `case` patterns, you'd just be checking using a type pattern or null-check pattern, and throw yourself in a default branch.

It might happen in the last case of the switch, where you know from invariants what is left, even if the types haven't been exhausted.
